### PR TITLE
Proof of Concept: Running Rust Crypto in Arbitrum

### PR DESCRIPTION
**Pull request by nomaxg**
This draft PR illustrates changes necessary to get Rust crypto code running in Arbitrum. Specifically it shows:
- How the JIT validator machine can call native Rust code directly (the machines are written in Rust)
- How to use cgo to run Rust from Go directly 
- How to compile Rust code to a wasm library that is linked to `replay.wasm`, the binary which is bisected during fraud proof arbitration

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -50,7 +50,7 @@ replay_wasm=$(output_root)/machines/latest/replay.wasm
 
 arbitrator_generated_header=$(output_root)/include/arbitrator.h
 arbitrator_wasm_libs_nogo=$(output_root)/machines/latest/wasi_stub.wasm $(output_root)/machines/latest/host_io.wasm $(output_root)/machines/latest/soft-float.wasm
-arbitrator_wasm_libs=$(arbitrator_wasm_libs_nogo) $(patsubst %,$(output_root)/machines/latest/%.wasm, go_stub brotli)
+arbitrator_wasm_libs=$(arbitrator_wasm_libs_nogo) $(patsubst %,$(output_root)/machines/latest/%.wasm, go_stub brotli vid)
```
Comment (nomaxg) [inline Makefile:53]: Include the new wasm-lib in the list of arbitrator libraries so that it is compiled to wasm

File: Makefile [hunk 2]
```diff
@@ -282,8 +282,13 @@ $(output_root)/machines/latest/brotli.wasm: $(DEP_PREDICATE) $(wildcard arbitrat
 	cargo build --manifest-path arbitrator/wasm-libraries/Cargo.toml --release --target wasm32-wasi --package brotli
 	install arbitrator/wasm-libraries/target/wasm32-wasi/release/brotli.wasm $@
 
+$(output_root)/machines/latest/vid.wasm: $(DEP_PREDICATE) $(wildcard arbitrator/wasm-libraries/vid/src/*)
```
Comment (nomaxg) [inline Makefile:285]: command to compile the vid code to wasm

File: Makefile [hunk 3]
```diff
@@ -282,8 +282,13 @@ $(output_root)/machines/latest/brotli.wasm: $(DEP_PREDICATE) $(wildcard arbitrat
...
+	mkdir -p $(output_root)/machines/latest
+	cargo build --manifest-path arbitrator/wasm-libraries/Cargo.toml --release --target wasm32-wasi --package vid
+	install arbitrator/wasm-libraries/target/wasm32-wasi/release/vid.wasm $@
+
 $(output_root)/machines/latest/machine.wavm.br: $(DEP_PREDICATE) $(arbitrator_prover_bin) $(arbitrator_wasm_libs) $(replay_wasm)
-	$(arbitrator_prover_bin) $(replay_wasm) --generate-binaries $(output_root)/machines/latest -l $(output_root)/machines/latest/soft-float.wasm -l $(output_root)/machines/latest/wasi_stub.wasm -l $(output_root)/machines/latest/go_stub.wasm -l $(output_root)/machines/latest/host_io.wasm -l $(output_root)/machines/latest/brotli.wasm
+	$(arbitrator_prover_bin) $(replay_wasm) --generate-binaries $(output_root)/machines/latest -l $(output_root)/machines/latest/soft-float.wasm -l $(output_root)/machines/latest/wasi_stub.wasm -l $(output_root)/machines/latest/go_stub.wasm -l $(output_root)/machines/latest/host_io.wasm -l $(output_root)/machines/latest/brotli.wasm -l $(output_root)/machines/latest/vid.wasm
```
Comment (nomaxg) [inline Makefile:291]: link the wasm binary to `replay.wasm`

File: arb-vid/Makefile [hunk 4]
```diff
@@ -0,0 +1,7 @@
+ROOT_DIR := $(dir $(realpath $(lastword $(MAKEFILE_LIST))))
+
+.PHONY: build-dynamic
+build-dynamic:
```
Comment (nomaxg) [inline arb-vid/Makefile:4]: make command to build the dynamic library used by cgo to call Rust from Go

File: arbitrator/jit/src/machine.rs [hunk 5]
```diff
@@ -119,6 +119,7 @@ pub fn create(opts: &Opts, env: WasmEnv) -> (Instance, FunctionEnv<WasmEnv>, Sto
 
             "github.com/offchainlabs/nitro/arbcompress.brotliCompress" => func!(arbcompress::brotli_compress),
             "github.com/offchainlabs/nitro/arbcompress.brotliDecompress" => func!(arbcompress::brotli_decompress),
+            "github.com/offchainlabs/nitro/arbcompress.verifyNamespace" => func!(arbcompress::verify_namespace),
```
Comment (nomaxg) [inline arbitrator/jit/src/machine.rs:122]: The JIT prover machine is written in Rust, so here we simply override the import and call the function natively

File: arbitrator/wasm-libraries/vid/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+//         encoded_size: *mut usize,
+//         encoded_buffer: *mut u8,
+//     ) -> u32;
+// }
+
+// const BROTLI_MODE_GENERIC: u32 = 0;
+// const BROTLI_RES_SUCCESS: u32 = 1;
+
+#[no_mangle]
+pub unsafe extern "C" fn go__github_com_offchainlabs_nitro_arbcompress_verifyNamespace(
```
Comment (nomaxg) [inline arbitrator/wasm-libraries/vid/src/lib.rs:31]: This weird naming convention allows the verifyNamespace function to be linked correctly after being compiled to wasm

File: cmd/replay/main.go [hunk 7]
```diff
@@ -154,6 +155,7 @@ func main() {
 	db := state.NewDatabase(raw)
 
 	lastBlockHash := wavmio.GetLastBlockHash()
+	arbcompress.CallTestVid()
```
Comment (nomaxg) [inline cmd/replay/main.go:158]: The entry point that wraps all of the messiness above

File: arbitrator/wasm-libraries/vid/src/lib.rs (hunk 6 above)
Comment (ImJeremyHe) [reply arbitrator/wasm-libraries/vid/src/lib.rs:31]: I think this is the convention of the go wasm tool

File: arbcompress/compress_wasm.go [hunk 8]
```diff
@@ -14,6 +14,8 @@ func brotliCompress(inBuf []byte, outBuf []byte, level int, windowSize int) int6
 
 func brotliDecompress(inBuf []byte, outBuf []byte) int64
 
+func verifyNamespace()
```
Comment (ImJeremyHe) [inline arbcompress/compress_wasm.go:17]: What is arbcompress? I have no idea about why this function is here

Comment (nomaxg) [reply arbcompress/compress_wasm.go:17]: I put it in a random place, when I clean this up I'll put it in its own arb-vid package

**Pull request closed without merging**